pennylane/operation.py

```diff
@@ -129,6 +129,101 @@
 
 from .utils import pauli_eigs
 
+
+def expand_matrix(base_matrix, wires, wire_order):
+    """Re-express a base matrix acting on a subspace defined by a set of wire labels
+    according to a global wire order.
+
+    This function is fully differentiable.
+
+    **Examples:**
+
+    If the wire order is identical to ``wires``, the original matrix gets returned:
+
+    >>> base_matrix = np.array([[1, 2, 3, 4],
+    >>>                         [5, 6, 7, 8],
+    >>>                         [9, 10, 11, 12],
+    >>>                         [13, 14, 15, 16]])
+    >>> expand_matrix(base_matrix, wires=[0, 2], wire_order=[0, 2])
+    [[ 1  2  3  4]
+     [ 5  6  7  8]
+     [ 9 10 11 12]
+     [13 14 15 16]]
+
+    If the wire order is a permutation of ``wires``, the entries of the base matrix get permuted:
+
+    >>> expand_matrix(base_matrix, wires=[0, 2], wire_order=[2, 0])
+    [[ 1  3  2  4]
+     [ 9 11 10 12]
+     [ 5  7  6  8]
+     [13 15 14 16]]
+
+    If the wire order contains wire labels not found in ``wires``, the matrix gets expanded:
+
+    >>> expand_matrix(base_matrix, wires=[0, 2], wire_order=[0, 1, 2])
+    [[ 1  2  0  0  3  4  0  0]
+     [ 5  6  0  0  7  8  0  0]
+     [ 0  0  1  2  0  0  3  4]
+     [ 0  0  5  6  0  0  7  8]
+     [ 9 10  0  0 11 12  0  0]
+     [13 14  0  0 15 16  0  0]
+     [ 0  0  9 10  0  0 11 12]
+     [ 0  0 13 14  0  0 15 16]]
+
+    Args:
+        base_matrix (tensor_like): base matrix to expand
+        wires (Iterable): wires determining the subspace that base matrix acts on; a base matrix of
+            dimension :math:`2^n` acts on a subspace of :math:`n` wires
+        wire_order (Iterable): global wire order, which has to contain all wire labels in ``wires``, but can also
+            contain additional labels
+    """
+
+    shp = qml.math.shape(base_matrix)
+    if shp[0] != shp[1]:
+        raise ValueError(f"Expected square matrix to expand; got shape {shp}.")
+
+    if 2 ** len(wires) != shp[0]:
+        raise ValueError(
+            f"Expected base matrix of size {2 ** len(wires)} to act on {len(wires)} wires; got size {dim}."
+        )
+
+    if wire_order is None:
+        return base_matrix
+
+    wire_order = Wires(wire_order)
+    wires = Wires(wires)
+
+    if Wires(wires) == Wires(wire_order):
+        return base_matrix
```

Comment:
I wonder if this should be moved into `self.matrix`:

```python
def matrix(self):
    base_matrix = self.compute_matrix(*self.parameters, **self.hyperparameters)

    if self.wires == Wires(wire_order):
        return base_matrix

    return expand_matrix(base_matrix, wires=self.wires, wire_order=wire_order)
```

the advantage is that it avoids a call to `expand_matrix` if not required

Reply:
Yes, I was wondering though that it has to be added every time we overwrite this method?

Reply:
Good point 🤔 My main motivation here is to avoid calling `expand_matrix` if not needed, to avoid (a probably minor) overhead, and to avoid additional validation that occurs that might break the JIT.

Reply:
I implemented the change, probably worth it!